x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/UnresolvedNamePattern.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.expression;
+
+import org.apache.lucene.util.automaton.CharacterRunAutomaton;
+import org.elasticsearch.xpack.ql.capabilities.UnresolvedException;
+import org.elasticsearch.xpack.ql.expression.Expression;
+import org.elasticsearch.xpack.ql.expression.Nullability;
+import org.elasticsearch.xpack.ql.expression.UnresolvedNamedExpression;
+import org.elasticsearch.xpack.ql.tree.NodeInfo;
+import org.elasticsearch.xpack.ql.tree.Source;
+import org.elasticsearch.xpack.ql.util.CollectionUtils;
+
+import java.util.List;
+import java.util.Objects;
+
+import static java.util.Collections.emptyList;
+
+/**
+ * Unresolved expression for encapsulating a pattern:
+ * KEEP `a*`, b*, `c*`*`d*`
+ * a* is an actual name (UnresolvedAttribute)
+ * b* is a name pattern (this class)
+ * `c*`*`d*` is a name pattern
+ */
+public class UnresolvedNamePattern extends UnresolvedNamedExpression {
+
+    private final CharacterRunAutomaton automaton;
+    private final String pattern;
+
+    public UnresolvedNamePattern(Source source, CharacterRunAutomaton automaton, String patternString) {
+        super(source, emptyList());
+        this.automaton = automaton;
+        this.pattern = patternString;
+    }
+
+    public boolean match(String string) {
+        return automaton.run(string);
+    }
+
+    @Override
+    public String name() {
+        return pattern;
+    }
+
+    public String pattern() {
+        return pattern;
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        throw new UnsupportedOperationException("this type of node doesn't have any children to replace");
+    }
+
+    @Override
+    protected NodeInfo<UnresolvedNamePattern> info() {
+        return NodeInfo.create(this, UnresolvedNamePattern::new, automaton, pattern);
+    }
+
+    @Override
+    public String unresolvedMessage() {
+        return "Unresolved pattern [" + pattern + "]";
+    }
+
+    public static String errorMessage(String pattern, List<String> potentialMatches) {
+        String msg = "No matches found for pattern [" + pattern + "]";
+        if (CollectionUtils.isEmpty(potentialMatches) == false) {
+            msg += ", did you mean to match "
```

Comment:
```suggestion
            msg += ", did you mean "
```

The message is slightly inconsistent with what we use in ql in general (`to match` is extra).

Reply:
I've done the change on purpose to signal it's a pattern not regular name matching.